### PR TITLE
document device list synchronisation over federation.

### PR DESCRIPTION
**Pull request by ara4n**
closes https://github.com/matrix-org/matrix-doc/issues/1212

### CODE REVIEW

File: api/server-server/definitions/event-schemas/m.device_list_update.yaml [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+# presence lists, profile info, etc).  But documenting what we have for now...
+
+allOf:
+  - $ref: ../edu.yaml
+  - type: object
+    properties:
+      edu_type:
+        type: enum
+        enum: ['m.device_list_update']
+        description: The string ``m.device_list_update``
```
Comment (turt2live) [inline api/server-server/definitions/event-schemas/m.device_list_update.yaml]: Descriptions should end in full stops.

File: api/server-server/definitions/event-schemas/m.device_list_update.yaml [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+            type: boolean
+            description: |-
+              True if the server is announcing that this device has been deleted.
+            example: false
+          keys:
+            description: |-
+              The updated identity keys (if any) for this device.  May be absent if the
+              device has no E2E keys defined.
+            allOf:
+              - $ref: ../../../client-server/definitions/device_keys.yaml
```
Comment (turt2live) [inline api/server-server/definitions/event-schemas/m.device_list_update.yaml]: I don't think this needs an `allOf` and can just have the `$ref` at the same level as `description`.

File: api/server-server/definitions/event-schemas/m.device_list_update.yaml (hunk 1 above)
Comment (turt2live) [reply api/server-server/definitions/event-schemas/m.device_list_update.yaml]: (this is a theme throughout)

File: specification/server_server_api.rst [hunk 3]
```diff
@@ -996,6 +996,46 @@ nothing else.
...
+Details of a user's devices must be efficiently published to other users and kept
+up-to-date.  This is critical for reliable end-to-end encryption, in order for users
+to know which devices are participating in a room.  It's also required for to-device
+messaging to work. This section is intended to complement the `Device Management module`_
+of the Client-Server API.
+
+Matrix currently uses a custom pubsub system for synchronising information
+about the list of devices for a given user over federation.  When a server
+wishes to determine a remote user's device list for the first time,
+it should populate its local cache by calling the /user/keys/query API
```
Comment (turt2live) [inline specification/server_server_api.rst]: preferably API routes should be surrounded in backticks (or be links)

File: specification/server_server_api.rst [hunk 4]
```diff
@@ -996,6 +996,46 @@ nothing else.
...
+for clustered servers which could send multiple EDUs at the same time, the ``prev_id``
+field should include all ``m.device_list_update`` EDUs which have not been yet been
+referenced in a EDU. If EDUs are emitted in series by a server, there should only ever
+be one ``prev_id`` in the EDU.
+
+This forms a simple directed acyclic graph of ``m.device_list_update`` EDUs, showing
+which EDUs a server needs to have received in order to apply an update to its local
+copy of the remote user's device list.  If a server receives an EDU which refers to
+a ``prev_id`` it does not recognise, it must resynchronise its list by calling the
+/user/keys/query API and resume the process.  The response contains a ``stream_id``
```
Comment (turt2live) [inline specification/server_server_api.rst]: backticks around api routes

**Comment by ara4n** [comment]
@erikjohnston please can you check this for factual accuracy so we can merge it (or at least fix & merge it? :)

**Comment by erikjohnston** [comment]
Ideally this would also talk about when a server should send device updates and to whom, as well as documenting what to do when a server receives an update that points to a prev stream ID it hasn't seen.

What's there looks broadly accurate.

**Comment by ara4n** [comment]
> Ideally this would also talk about when a server should send device updates and to whom

Yup, it didn't spell out who the device updates should be sent to; hopefully now fixed.

> as well as documenting what to do when a server receives an update that points to a prev stream ID it hasn't seen.

This was documented twice:

> If a server receives an EDU which refers to a ``prev_id`` it does not recognise, it must resynchronise its list by calling the ``/user/keys/query API`` and resume the process.

in the s2s verbiage

and

> If the receiving server does not recognise any of the prev_ids, it means an EDU has been lost and the server should query a snapshot of the devicelist via ``/user/keys/query`` in order to correctly interpret future ``m.device_list_update`` EDUs.

in the m.device_list_update definition.

If this is insufficient, can you please say what the gap is so I can fill it in?

**Comment by turt2live** [comment]
Finishes off the last of https://github.com/matrix-org/matrix-doc/issues/501

File: specification/server_server_api.rst [hunk 5]
```diff
@@ -996,6 +996,50 @@ nothing else.
...
+about the list of devices for a given user over federation.  When a server
+wishes to determine a remote user's device list for the first time,
+it should populate a local cache from the result of a ``/user/keys/query`` API
+on the remote server.  However, subsequent updates to the cache should be applied
+by consuming ``m.device_list_update`` EDUs.  Each new ``m.device_list_update`` EDU
+describes an incremental change to one device for a given user which should replace
+any existing entry in the local server's cache of that device list. Servers must send
+``m.device_list_update`` EDUs to all the servers whose users participate in their rooms,
+and must be sent whenever a local user's device list changes (i.e. new or deleted devices,
+or changes of identity keys).
```
Comment (erikjohnston) [inline specification/server_server_api.rst]: Or a user joins (gets invited?) into a room.

**Comment by erikjohnston** [comment]
D'oh. Apparently you need to say it a few more times for it to actually sink here.

Otherwise LGTM modulo small comment

**Comment by ara4n** [comment]
hopefully fixed the small comment.

**Pull request merged**